### PR TITLE
Ft6 proxy loss refactor

### LINKED ISSUE #9: Create a class hierarchy for proxy-loss types: Quadratic/XEntropy/Hinge
## Summary

- Currently we have to do string comparisons multiple times per data row, per iteration, to figure out which proxy loss function to use to compute the gradients.
  - e.g., https://github.com/feedzai/fairgbm/blob/bulk-fairgbm-modifications/src/objective/constrained_recall_objective.hpp#L86
- According to our benchmarks, this can take up around 30% of the total training time.
- We should create a common interface w/ different subclasses to optimize this.
- This has already been started in #18

### PR DESCRIPTION
**Pull request by sgpjesus**
Closes #9

### CODE REVIEW

File: include/LightGBM/objective_function.h [hunk 1]
```diff
@@ -9,6 +9,11 @@
 #include <LightGBM/dataset.h>
 #include <LightGBM/meta.h>
 #include <LightGBM/utils/constrained.hpp>
+#include <LightGBM/proxy-loss-types/proxy_loss_base.hpp>
```
Comment (AlbertoEAF) [inline include/LightGBM/objective_function.h]: 1. Do not use "-" in folder names, replace with "_"'s.
2. We should not be adding more stuff to objective_function.h but instead to constrained_objective_function.h? Or this becomes unmaintainable when we try to merge newer patches from LGBM

File: include/LightGBM/objective_function.h [hunk 2]
```diff
@@ -95,11 +100,13 @@ class ObjectiveFunction {
   LIGHTGBM_EXPORT static ObjectiveFunction* CreateObjectiveFunction(const std::string& str);
 };
 
+namespace Constrained {
```
Comment (AlbertoEAF) [inline include/LightGBM/objective_function.h]: Haven't we split yet the constrained objective function code from the objective_function.h/.cpp into a separate constrained_objective_function.h/.cpp?

I though we already had done that

Failing to put all this new code on separate new files will make maintenance where new patches are merged from LGBM impossible.

Comment (AndreFCruz) [reply include/LightGBM/objective_function.h]: we did and it was reverted. I've un-reverted it in the last commit :slightly_smiling_face:

**Comment by AndreFCruz** [comment]
I'm happy to report that the training time on the AOF dataset (~750k samples) has improved from ~3m30s to ~2m5s with this refactor :tada: :tada: 

I've checked LightGBM and FairGBM predictions against reference predictions (both from Microsoft's `master` branch and our `main-fairgbm` branch), and everything is OK :slightly_smiling_face: 

Notebook used to test this PR: 
[test-fairgbm-feat6.pdf](https://github.com/feedzai/fairgbm/files/9498132/test-fairgbm-feat6.pdf)


(we should also use this notebook to create similar system tests on the python package)
